src/LazyHydrate.js

```diff
@@ -0,0 +1,141 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+
+import { StylesProvider, createGenerateClassName } from '@material-ui/core/styles'
+import { SheetsRegistry } from 'jss'
+
+const registries = []
+const generateClassName = createGenerateClassName()
+
+/*
+  This component renders the server side rendered stylesheets for the
+  lazy hydrated components. Once they become hydrated, these stylesheets
+  will be removed.
+*/
+export function LazyStyleElements() {
+  return (
+    <>
+      {registries.map((registry, index) => {
+        const id = `jss-lazy-${index}`
+        return <style key={id} id={id} dangerouslySetInnerHTML={{ __html: registry.toString() }} />
+      })}
+    </>
+  )
+}
+
+function LazyStylesProvider({ children }) {
+  const registry = new SheetsRegistry()
+  registries.push(registry)
+  return (
+    <StylesProvider
+      sheetsManager={new Map()}
+      serverGenerateClassName={generateClassName}
+      sheetsRegistry={registry}
+    >
+      {children}
+    </StylesProvider>
+  )
+}
+
+const isBrowser =
+  typeof window !== 'undefined' &&
+  typeof window.document !== 'undefined' &&
+  typeof window.document.createElement !== 'undefined'
+
+// Used for detecting when the wrapped component becomes visible
+const io =
+  isBrowser && IntersectionObserver
+    ? new IntersectionObserver(entries => {
+        entries.forEach(entry => {
+          if (entry.isIntersecting || entry.intersectionRatio > 0) {
+            entry.target.dispatchEvent(new CustomEvent('visible'))
+          }
+        })
+      })
+    : null
+
+function LazyHydrateInstance({ hydrated, ssrOnly, children, on, index, ...rest }) {
+  const childRef = React.useRef(null)
+  const [_hydrated, setHydrated] = React.useState(!isBrowser)
+
+  React.useEffect(() => {
+    if (ssrOnly || hydrated) return
+
+    function hydrate() {
+      setHydrated(true)
+      // Remove the server side generated stylesheet
+      const stylesheet = window.document.getElementById(`jss-lazy-${index}`)
+      if (stylesheet) {
+        stylesheet.remove()
+      }
+    }
+
+    let el
+    if (on === 'visible') {
+      if (io && childRef.current.childElementCount) {
+        // As root node does not have any box model, it cannot intersect.
+        el = childRef.current.children[0]
+        io.observe(el)
+      }
+    }
+
+    childRef.current.addEventListener(on, hydrate, {
+      once: true,
+      capture: true,
+      passive: true,
+    })
+
+    return () => {
+      if (el) io.unobserve(el)
+      childRef.current.removeEventListener(on, hydrate)
+    }
+  }, [])
```

Comment:
```suggestion
  }, [ssrOnly, hydrated])
```

Reply:
I only want this function to run once. That is why I used an empty array.

Reply:
What's the use case for the `hydrated` prop? It's not documented nor used internally.
This is more about following best practices. As a rule of thumb hooks should never have to lie about their dependencies. The behavior of the effect is different based on the value of the `ssrOnly` prop. If one were to tie the value passed to `ssrOnly` to changing state for example, then the component would not work as expected. In 99% of cases the value of `ssrOnly` passed will be static, meaning `[ssrOnly]` and `[]` as dependencies will be equivalent - the effect is run once. But for cases where the integrator wants to switch from `ssrOnly={true}` to `ssrOnly={false}`, the component should still work as expected. L62 should also be changed to `if (ssrOnly || _hydrated) return` so that the effect can be rerun multiple times while only hydrating once.

Reply:
`on` is also an important dependency. It would be a valid use case to switch from `on="click"` to `on="visible"` for example. That would not work if the dependency isn't specified.